### PR TITLE
translate snodgrass-99 tests to Datalog

### PR DESCRIPTION
**Pull request by refset**

### CODE REVIEW

File: src/test/clojure/core2/core/datalog_test.clj [hunk 1]
```diff
@@ -1012,3 +1012,197 @@
...
+      (t/is (= [{:cust 827, :app-start (util/->zdt #inst "1998-01-12"), :app-end (util/->zdt #inst "1998-01-20")}]
+               (q '{:find [cust app-start app-end]
+                    :order-by [[app-start :asc]]
+                    :where [(for-all-app-time id)
+                            [id :application_time_start app-start]
+                            [id :application_time_end app-end]
+                            [id :customer-number cust]]}, !tx6, nil))
+            "'as best known' (as-of 30 Jan)")
+
+      (t/is (= [{:prop 3621}]
```
Comment (refset) [inline src/test/clojure/core2/core/datalog_test.clj:1103]: Note that this test only returns 1 result, which means there is some unresolved semantic drift with what I cooked up previously:
https://github.com/xtdb/core2/blob/31255227aa1b3c3e2d104da8a2b13310b7124503/src/test/resources/core2/sql/logic_test/direct-sql/sl-demo.test#L677-L693

The original from the textbook is (and presumably only returns one result but I'm not sure that can be confirmed):
```sql
What properties are owned by the customer who owns property
7797?
SELECT P2.property_number
FROM Prop_Owner AS P1, Prop_Owner AS P2
WHERE P1.property_number = 7797
AND P2.property_number <> P1.property_number
AND P1.customer_number = P2.customer_number
```

...I haven't dug into the logical plans for these, but I reckon the top-level semantics are sufficiently different that the right answer was probably to add a `distinct` in the sl-demo version in the first instance.

**Review by jarohen — APPROVED** [review]
@refset: fine with me, once you're happy with the 1 vs 3 rows feel free to merge :slightly_smiling_face:

**Comment by refset** [comment]
From the book:
```sql
```
```clojure
[:rename
 {x4 property_number}
 [:project
  [x4]
  [:mega-join
   [(<> x4 x1) {x2 x5}]
   [[:rename
     {property_number x1, customer_number x2}
     [:scan
      prop_owner
      [{property_number (= property_number 7797)} customer_number]]]
    [:rename
     {property_number x4, customer_number x5}
     [:scan prop_owner [property_number customer_number]]]]]]]
```

Datalog:
```
[:project
 [prop]
 [:order-by
  [[app-start {:direction :asc}]]
  [:select
   (<> prop in-prop)
   [:project
    [{sys-end _r0_sys-end}
     {in-prop _r1_in-prop}
     {app-end _r0_app-end}
     {id2 _r0_id2}
     {sys-start _r0_sys-start}
     {cust _r0_cust}
     {id _r1_id}
     {prop _r0_prop}
     {app-start _r0_app-start}]
    [:select
     (and (= _r1_in-prop ?in-prop) (= _r0_cust _r1_cust))
     [:mega-join
      []
      [[:rename
        {sys-end _r0_sys-end,
         app-end _r0_app-end,
         id2 _r0_id2,
         sys-start _r0_sys-start,
         cust _r0_cust,
         prop _r0_prop,
         app-start _r0_app-start}
        [:project
         [{cust customer-number}
          {prop property-number}
          {sys-start system_time_start}
          {sys-end system_time_end}
          {app-start application_time_start}
          {app-end application_time_end}
          {id2 id}]
         [:scan
          xt_docs
          [system_time_end
           application_time_start
           application_time_end
           customer-number
           property-number
           id
           system_time_start]]]]
       [:rename
        {in-prop _r1_in-prop, cust _r1_cust, id _r1_id}
        [:project
         [{in-prop property-number} {cust customer-number} id]
         [:scan xt_docs [customer-number property-number id]]]]]]]]]]]
```

**Comment by refset** [comment]
side-by-side, Datalog on the left, SQL on right
image

**Comment by refset** [comment]
Datalog post rewrites:
```clojure
[:project
 [prop]
 [:order-by
  [[app-start {:direction :asc}]]
  [:select
   (<> prop in-prop)
   [:project
    [{id2 _r0_id2}
     {cust _r0_cust}
     {prop _r0_prop}
     {in-prop _r1_in-prop}
     {id _r1_id}]
    [:select
     (and (= _r0_cust _r1_cust) (= _r1_in-prop ?in-prop))
     [:mega-join
      []
      [[:rename
        {id2 _r0_id2, cust _r0_cust, prop _r0_prop}
        [:project
         [{cust customer-number} {prop property-number} {id2 id}]
         [:scan xt_docs [customer-number property-number id]]]]
       [:rename
        {in-prop _r1_in-prop, cust _r1_cust, id _r1_id}
        [:project
         [{in-prop property-number} {cust customer-number} id]
         [:scan xt_docs [customer-number property-number id]]]]]]]]]]]
```

**Comment by refset** [comment]
Blocked by #655

**Comment by refset** [comment]
For the record, the query returning 3 items wasn't strictly valid anyway due to the textbook example being based on a hypothetical "snapshot table" - so I've removed it completely.

**Pull request merged**